bonfire/bonfire.py

```diff
@@ -81,6 +89,51 @@ def current_namespace_or_error():
     return namespace
 
 
+def option_usage_wrapper(command):
+    def decorator(f):
+        @wraps(f)
+        def wrapper(*args, **kwargs):
+            # only gather telemetry if enabled and a bot run
+            options_used = []
+            is_parameter = False
+            for arg in sys.argv[2:]:
+                if is_parameter:
+                    options_used.append(arg.split('=')[0])
+                    is_parameter = False
+                elif arg == '-p' or arg == '--set-parameter':
+                    is_parameter = True
+
+            start_time = time.time()
+
+            log_entry = {
+                "timestamp": datetime.datetime.now().isoformat(),
+                "duration_seconds": time.time() - start_time,
+                "bot": conf.BONFIRE_BOT,
+                "command": command,
```

Comment:
What if we move some of this data right into the ` AsyncElasticsearchHandler` class?

When we initialize the instance at the startup of bonfire (in `main_with_handler`), we can give that instance some a 'metadata' variable. So for example, the class could look like this:

```
class AsyncElasticsearchHandler(logging.Handler):
    def __init__(self, es_url, metadata):
        super().__init__()
        self.es_url = es_url
        self.executor = ThreadPoolExecutor(max_workers=10)
        self.metadata = metadata
```

and in `main_with_handler` we could do something like:
```python
metadata = {
    "uuid": uuid.uuid4()
    "start_time": datetime.datetime.now().isoformat(),
    "bot": conf.BONFIRE_BOT,
    "command": _mask_parameter_values(sys.argv[1:]),  # need to write code for  _mask_parameter_values
}
es_handler = AsyncElasticsearchHandler(conf.ELASTICSEARCH_HOST, metadata)
```

and every time that `emit` is called, it could add this `metadata` into the JSON data that it sends? This could be considered the `run metadata` or something that never changes for the entire `bonfire` run until the process exits. As long as the `uuid` stays the same for the whole process execution -- it would allow us to track things for the same run of bonfire.

You may be able to have some helper funcs in the class itself that add on the `success` or the `failure` data as well.

Basically the end goal here is: I think we want to be able to call `send_telementry("something happened...")` at more parts of the code in future... kind of like how we might call `log.info("something happened...") at various places. So I am thinking we want some helper functions defined that can be called in various pieces of code. When we call them, the "run ID" that is sent to ES should stay the same ... even though we might send a few telemetry messages for 1 bonfire command execution.

Reply:
I added a couple helper functions to get this working the way you describe. I'm not exactly sure where these helper functions should live. Right now I have them in `bonfire.py` but I may move them to `elastic_logging.py` to keep all the related functions together. 
The only reason I didn't do that yet is because it will make things a bit more complex. ie. making sure a logger exists. If I design it so the handler stores a reference to the logger and just call the logger from the handler, well that's just weird I guess. Having the handler actually be the logger.
There may be a better solution, I'm just still brainstorming on what that is

Reply:
Ok I moved some pieces around so that the logic for elasticsearch is only in `elastic_logging.py`. Now you would just create an instance of `ElasticLogger` at the top of a file and call the object's `send_telemetry()`.